python_modules/dagster/dagster/_core/definitions/time_window_partitions.py

```diff
@@ -128,15 +128,79 @@ def __new__(  # pylint: disable=arguments-differ
             cls, start_dt, timezone or "UTC", fmt, end_offset, cron_schedule
         )
 
-    def get_partitions(
-        self, current_time: Optional[datetime] = None
-    ) -> Sequence[Partition[TimeWindow]]:
-        current_timestamp = (
+    def get_current_timestamp(self, current_time: Optional[datetime] = None) -> float:
+        # return pendulum.now(self.timezone).timestamp()
+        return (
             pendulum.instance(current_time, tz=self.timezone)
             if current_time
             else pendulum.now(self.timezone)
         ).timestamp()
 
+    def get_num_partitions(self, current_time: Optional[datetime] = None) -> int:
+        # Method added for performance reasons.
+        # Fetching partition keys requires significantly more compute time to
+        # string format datetimes.
+        current_timestamp = self.get_current_timestamp(current_time=current_time)
+
+        partitions_past_current_time = 0
+
+        num_partitions = 0
+        for time_window in self._iterate_time_windows(self.start):
+            if (
+                time_window.end.timestamp() <= current_timestamp
+                or partitions_past_current_time < self.end_offset
+            ):
+                num_partitions += 1
+
+                if time_window.end.timestamp() > current_timestamp:
+                    partitions_past_current_time += 1
+            else:
+                break
+
+        if self.end_offset < 0:
+            num_partitions += self.end_offset
+
+        return num_partitions
```

Comment:
Man it's really tough that we support full-complexity cron strings here 😅  This algorithm looks good to me knowing that we really have no choice but to walk through time.

In the UI we're going to request this for every asset in the asset graph along with other "live" asset info -- I think this is probably performant enough but if we find that it's not, I wonder if we could add some light caching around this function call. (maybe even cache the result in an instance var here?)

 eg: If I ask for the "num partitions" of 50 assets and all 50 of them reference the same partition definition, could we just count partitions in that definition once during the GraphQL request?